src/text/parsers/decimal.rs

```diff
@@ -0,0 +1,211 @@
+    let sign = if sign_text.is_some() {Sign::Negative} else {Sign::Positive};
+
+    // TODO: Reusable buffer for formatting/sanitization
+    let mut magnitude_text = format!("{}{}", digits_before_dot, digits_after_dot);
+    magnitude_text.retain(|c| c != '_');
+    // Ion's parsing rules should only let through strings of digits and underscores. Since
+    // we've just removed the underscores above, the `from_str` methods below should always
+    // succeed.
+    let magnitude: Magnitude = if magnitude_text.len() < MAX_U64_DIGITS {
+        u64::from_str(&magnitude_text)
+            .expect("parsing coefficient magnitude as u64 failed")
+            .into()
+    } else {
+        BigUint::from_str(&magnitude_text)
+            .expect("parsing coefficient magnitude as BigUint failed")
+            .into()
+    };
+
+    let coefficient = Coefficient::new(sign, magnitude);
+
+    // TODO: Reusable buffer for sanitization
+    let sanitized = exponent_text.replace('_', "");
+    let mut exponent = i64::from_str(&sanitized)
+        .expect("parsing exponent as i64 failed");
+    // Reduce the exponent by the number of digits that follow the decimal point
+    exponent -= digits_after_dot.chars().filter(|c| c.is_digit(10)).count() as i64;
+    Decimal::new(coefficient, exponent)
+}
+
+/// Matches the decimal exponent marker ('d' or 'D') followed by a signed integer. (e.g. 'd-16')
+fn decimal_exponent_marker_followed_by_digits(input: &str) -> IResult<&str, &str> {
+    preceded(
+        one_of("dD"),
+        exponent_digits
+    )(input)
+}
+
+#[cfg(test)]
+mod reader_tests {
+    use crate::text::parsers::decimal::parse_decimal;
+    use crate::text::parsers::unit_test_support::{parse_test_err, parse_test_ok};
+    use crate::text::TextStreamItem;
+    use crate::types::decimal::Decimal;
+
+    fn parse_equals(text: &str, expected: Decimal) {
+        parse_test_ok(parse_decimal, text, TextStreamItem::Decimal(expected))
+    }
+
+    fn parse_fails(text: &str) {
+        parse_test_err(parse_decimal, text)
+    }
+
+    #[test]
+    fn test_parse_decimals_with_exponents() {
+        parse_equals("0d0 ", Decimal::new(0, 0));
+        parse_equals("0D0 ", Decimal::new( 0, 0));
+        parse_equals("-0d0 ", Decimal::negative_zero());
+        parse_equals("-0D0 ", Decimal::negative_zero());
+        parse_equals("-0d5 ", Decimal::negative_zero_with_exponent(5));
+        parse_equals("-0d-5 ", Decimal::negative_zero_with_exponent(-5));
+        parse_equals("305d1 ", Decimal::new(305, 1));
+        parse_equals("305d-1 ", Decimal::new( 305, -1));
+        parse_equals("111_111d222 ", Decimal::new(111_111, 222));
+        parse_equals("111_111d-222 ", Decimal::new(111_111, -222));
+        parse_equals("111_111d222_222 ", Decimal::new(111_111, 222_222));
+        parse_equals("-279d0 ", Decimal::new(-279, 0));
+        parse_equals("-999_9d9_9 ", Decimal::new(-9_999, 99));
+        parse_equals("-999_9d-9_9 ", Decimal::new(-9_999, -99));
```

Comment:
Yes, Ion's text encoding requires that many of the scalar types end in a stop character. Here's an excerpt from [the spec section on Real Numbers](https://amzn.github.io/ion-docs/docs/spec.html#real-numbers):
> In the text notation, real values must be followed by one of the fifteen numeric stop-characters: `{}[](),\"\'\ \t\n\r\v\f`.

(It's tough to see in the font above but one of those characters is a space.)

Additionally, `nom` will reject any input that might be incomplete data. For example, the text `314` might be a complete integer or it might be the first half of a `314159`. Adding a trailing space removes this ambiguity.

Reply:
Thanks. I forgot the Ion spec.